dataflow/flex-templates/pipeline_with_dependencies/README.md

```diff
@@ -0,0 +1,169 @@
+Using the runtime image as the base image allows us to reduce the number of Docker images that need to be maintained.
+It also ensures that the pipeline uses the same dependencies at submission and at runtime.
+
+```sh
+export TEMPLATE_FILE=gs://$BUCKET/longest-word-$TAG.json
+export TEMPLATE_IMAGE=$REGION-docker.pkg.dev/$PROJECT/$REPOSITORY/my_template_image:$TAG
+```
+
+```sh
+gcloud dataflow flex-template build $TEMPLATE_FILE  \
+    --image-gcr-path $TEMPLATE_IMAGE \
+    --sdk-language "PYTHON" \
+    --flex-template-base-image $SDK_CONTAINER_IMAGE \
+    --py-path "."  \
+    --env "FLEX_TEMPLATE_PYTHON_PY_FILE=main.py" \
+    --env "FLEX_TEMPLATE_PYTHON_SETUP_FILE=setup.py" \
+    --project $PROJECT
+```
+
+Note: Older versions of `gcloud` might not support Artifact Registry in the `--flex-template-base-image` option. You might need to upgrade `gcloud` or build the Flex Template from a `Dockerfile` using the instructions below.
+
+If you prefer to have a separate file to configure and build the Flex Template image (included in the example), such as a `Dockefile.flex_template` file, you can build the Flex Template as follows:
+
+```sh
+gcloud builds submit --config flex_template_cloudbuild.yaml \
+    --substitutions="_TEMPLATE_IMAGE=${TEMPLATE_IMAGE},_BASE_IMAGE=${SDK_CONTAINER_IMAGE}" \
+    --project $PROJECT
+
+gcloud dataflow flex-template build $TEMPLATE_FILE  \
+    --image $TEMPLATE_IMAGE \
+    --sdk-language "PYTHON" \
+    --project $PROJECT
+```
+
+## Run the template
+
+```sh
+gcloud dataflow flex-template run "flex-`date +%Y%m%d-%H%M%S`" \
+    --template-file-gcs-location $TEMPLATE_FILE \
+    --region $REGION \
+    --staging-location "gs://$BUCKET/staging" \
+    --parameters input="gs://dataflow-samples/shakespeare/hamlet.txt" \
+    --parameters output="gs://$BUCKET/output" \
+    --parameters sdk_container_image=$SDK_CONTAINER_IMAGE \
+    --project $PROJECT
+```
+
+After the pipeline finishes, use the following command to inspect the output:
+```
+gsutil cat gs://$BUCKET/output*
+```
+
+## Optional: Update the dependencies in requirements.txt and rebuild the Docker images
+
+The top-level pipeline dependencies are defined in the `install_requires` section of the `setup.py` file.
+
+The `requirements.txt` file pins all Python dependencies, including transitive dependencies, that must be installed in the Docker base images. This step produces a reproducible set of dependencies every time the image is built.
+Version control the `requirements.txt` file together with the rest of pipeline code.
+
+When the dependencies of your pipeline change or when you want to use the latest available versions of packages in the pipeline's dependency chain,  regenerate the `requirements.txt` file:
+
+```
+    python3.11 -m pip install pip-tools   # Use a consistent minor version of Python throughout the project.
```

Comment:
Do we need to pin python to 3.11? shall we just do `python3`?

Reply:
pinning is better since dependency resolution can differ between minor versions. it can also be sensitive to which OS is being used...

Reply:
example: numpy is no longer released on some earlier versions of python, so it will resolve to an earlier version.